Style fixes & alternate to PR #2

zbeekman opened the pull request:
I've split this PR over three commits.

- The first is mostly white-space changes. Most of the changes can be made via `brew style --fix wrf`
- The second is to address audit failures. Again, `brew audit --strict --fix wrf` will do most of the work for you
- Lastly use a non-stable git repo & branch & sha as the version until there is a new stable release so that users can build with clang.

This is a proposed alternative to PR #2. You're welcome to ignore it though. This should resolve #1, in that it does things the canonical homebrew way, and enables compilation with or without `--HEAD`. (But it doesn't switch the C compiler to GCC). To do that a user could do `brew install --cc=gcc-9 --build-from-source wrf` if they so desired, although I wouldn't recommend it to them without a good reason.

zbeekman committed:
wrf: Cosmetic changes to conform with brew style

 Run `brew style --fix wrf` to fix nearly all of it

zbeekman committed:
wrf: Fix `brew audit --strict` failures

zbeekman committed:
wrf: Newer version to enable clang support

 - The versioning will need to be reverted/updated once a new stable release is
   out.

zbeekman commented:
https://github.com/WRF-CMake/homebrew-wrf/pull/4/files?w=1 should show the diff w/o whitespace changes.

zbeekman commented on wrf.rb:
```diff
@@ -2,66 +2,68 @@
+  url "https://github.com/WRF-CMake/WRF.git",
+      :branch => "wrf-cmake",
+      :commit => "8b9edc944f0cbbfb22966850ae3724f617f0f705"
+  version "4.1.wrf-cmake"
```
You may want to change this. I didn't try running `git describe` on the `wrf-cmake` branch, however that could be a descriptive candidate for a version string.

Once you have a stable release that allows clang, you'll probably want to switch back to specifying the release tarball.

letmaik replied:
Sure, looks fine.

zbeekman commented on wrf.rb:
```diff
@@ -2,66 +2,68 @@
+      :commit => "8b9edc944f0cbbfb22966850ae3724f617f0f705"
+  version "4.1.wrf-cmake"
+  sha256 "37814ee7bfe7077cba8bd0175258ef763380fce3e7d8aab2ea8360a902a11043"
+  head "https://github.com/WRF-CMake/WRF.git", :branch => "wrf-cmake"
```
if `wrf-cmake` is defined as the default branch of the repository, you probably don't need the branch specification, although it isn't hurting anyone.

letmaik replied:
Yep, I would leave it in to be more explicit.

letmaik commented on wrf.rb:
```diff
@@ -2,66 +2,68 @@
+  depends_on "open-mpi"
+
+  resource "WPS" do
+    url "https://github.com/WRF-CMake/WPS/archive/WPS-CMake-4.0.2.tar.gz"
```
This also needs Clang support, so would have to switch to `wps-cmake` branch as well. (I merged https://github.com/WRF-CMake/WPS/pull/13 now to allow this)

zbeekman replied:
Done in https://github.com/WRF-CMake/homebrew-wrf/pull/4/commits/7350c59120e33c60f46c164cdcfea7d31d33fc8c

zbeekman committed:
wrf: Update vendored WPS

 - Ensure a recent enough commit of WPS is being used that includes changes to
   allow compilation with clang.

zbeekman committed:
wrf: Enforce gfortran as the Fortran compiler

 - This prevents CMake from being promiscuous and finding Intel's `ifort`

zbeekman committed:
Rename formula to wrf-cmake

zbeekman committed:
wrf-cmake: install into prefix/"wrf"

 - Now wrf and wps are sibling directories in the install prefix

zbeekman commented:
FYI, running `brew linkage` shows the recursive linkage of the formula.

```
System libraries:
  /usr/lib/libSystem.B.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /usr/local/lib/gcc/9/libgcc_s.1.dylib (gcc)
  /usr/local/opt/gcc/lib/gcc/9/libgfortran.5.dylib (gcc)
  /usr/local/opt/gcc/lib/gcc/9/libquadmath.0.dylib (gcc)
  /usr/local/opt/jasper/lib/libjasper.4.dylib (jasper)
  /usr/local/opt/jpeg/lib/libjpeg.9.dylib (jpeg)
  /usr/local/opt/libpng/lib/libpng16.16.dylib (libpng)
  /usr/local/opt/netcdf/lib/libnetcdf.15.dylib (netcdf)
  /usr/local/opt/netcdf/lib/libnetcdff.6.dylib (netcdf)
  /usr/local/opt/open-mpi/lib/libmpi.40.dylib (open-mpi)
  /usr/local/opt/open-mpi/lib/libmpi_mpifh.40.dylib (open-mpi)
  /usr/local/opt/open-mpi/lib/libmpi_usempi_ignore_tkr.40.dylib (open-mpi)
  /usr/local/opt/open-mpi/lib/libmpi_usempif08.40.dylib (open-mpi)
Indirect dependencies with linkage:
  jpeg
```

Related to our discussion in #5, I'm not sure there are any greater risks in binary distribution than building from source against rolling-release dependencies. Our experience with homebrew has shown the opposite to be true: the builds are more predictable and the software shipped is less buggy when built in a very controlled environment. (We have anonymous analytics about errors encountered with any given formula, that can easily be opted out of.) As an example, @dmey's Intel compiler issue would not have occurred with a bottle because the CI machines are carefully provisioned and don't have unexpected user modifications. The protections in place beyond consistent build environment include:

1. Running the `test do` block after any (recursive) dependency is updated
2. Checking the linkage is correct for all (recursive) dependencies when any one changes

I'm not arguing here; your concerns are definitely still valid ones. I just wanted to point out a few of the potential benefits of shipping binaries. You can set up bottle distribution for a private tap too: that's what brewsci/homebrew-bio and brewsci/homebrew-num do. At any rate running `brew linkage wrf-cmake` after I built the formula made me realized I still had a brewed MPICH linked so I rebuilt after unlinking mpich and linking open-mpi.

letmaik commented:
I agree with you that the risks when compared to building from source via brew are similar. I guess it's more a matter of making people aware of the potential risks with Homebrew builds in general, compared to either manual builds (where people have full control over compilers and library versions) or using our validated pre-built binaries.

I don't fully understand your issue with MPICH linkage. Are you saying your installation was somehow in an inconsistent state? Doesn't homebrew prevent such things?

letmaik commented on wrf-cmake.rb:
```diff
@@ -0,0 +1,73 @@
+    ENV["FC"] = Formula["gcc"].opt_bin/"gfortran"
+
+    wrf_args = std_cmake_args.map do |s|
+      s.sub /-DCMAKE_INSTALL_PREFIX.*$/, "-DCMAKE_INSTALL_PREFIX=#{prefix}/wrf"
```
Why not use the approach I used below for WPS? Overriding works just fine and looks far simpler.

zbeekman replied:
Yes, that works. I didn't realize you were also passing `std_cmake_args` for WPS. I'm happy to do that, but it does rely on the behavior of CMake that CMake definitions on the right of the command line override those on the left. (`-DCMAKE_INSTALL_PREFIX` is being passed twice to CMake) I'm more confident of the stability of Ruby syntax than I am of the semantics of CMake command line defines. At the end of the day both implementations are effectively the same, but I agree that they should probably match. I'll go ahead and change this to match WPS, since it sounds like you prefer that.

zbeekman commented:
> I don't fully understand your issue with MPICH linkage. Are you saying your installation was somehow in an inconsistent state? Doesn't homebrew prevent such things?

I was doing sketchy things with modified code that I'm working on for Homebrew. Any build on CI will 100% avoid this. If a user builds from source they might be doing sketchy things too. This is partially caused by the fact that Homebrew doesn't sanitize the CMake build environment quite as well as I would like. My top # 1 and # 2 projects for Homebrew right now are:

1. Improving Fortran support so that:
   1. It won't go finding the Intel Compiler
   2. It will have the same, or similar shim and logic for compiler choice as for C and C++ and if a user builds from source with `-cc=gcc-#` the corresponding GFortran will be used for compilation rather than the latest stable version.
2. Don't set `CMAKE_PREFIX_PATH` to `/usr/local`, instead build up a list of recursive dependencies from `/usr/local/opt/<formula>`

Number 1 will solve issues like the one @dmey experienced, and help prevent me from doing sketchy things with MPICH, and 2 will ensure that accidental linking to a different implementation of a required library found via a FindModule won't happen.

zbeekman commented on wrf-cmake.rb:
```diff
@@ -0,0 +1,73 @@
+    # of WRF.
+    wps_args = std_cmake_args + %W[
+      -DCMAKE_INSTALL_PREFIX=#{prefix}/wps
+      -DWRF_DIR=#{buildpath}/build
```
I'm curious: Why is this necessary? What needs to be found from the build directory? Are you linking in a library target or something?

letmaik replied:
Yes, static libraries. WPS depends on WRF. We don't want to support the use case of really installing WRF as a first-class library, since this would be something new.

zbeekman replied:
I don't really get why WPS is a separate package then. It feels like the worst of both worlds. But I understand people are slow to adopt new things, and you may need to gently coax them into the 21st century... So if you think this is necessary (for now at least) to increase the odds of upstream accepting this improvement, then I'm all for it.

An idea for splitting the homebrew formula would be to have the wrf formula copy the needed libraries into a directory in the install tree (like `libs_for_wps`) then point the WPS install to that directory.

IMO, splitting out WPS can probably be tabled due to the unusual installation tree and strange (but less strange than upstream, 😄) build practices. Hopefully in the future options in CMake can be added to start:

* using normal binary names
* installing to canonical locations
* installing libraries needed by dependent packages
* install a cmake config file

Have you been in touch with upstream throughout this process? It's possible that they may either a) be completely un-receptive to change or b) be more willing than you realize to modernize things. It's hard to read minds. But again, I defer to you and your judgement, you're much more familiar with the community than I am.

letmaik replied:
I fully agree with your thinking. Currently we are conservative to have it as a compatible drop-in replacement. In the future this may change, and we will probably make suggestions to upstream about that, but given that WRF is decades old, it takes time. We've been in touch with upstream but interest was still low at that time. We'll re-engage once the paper is out and we have a bit more momentum.
I think all your changes are good to go now. Thanks again for helping us out here and being so patient! The world needs more people like you :)

zbeekman replied:
It would be pretty straight forwards to have a compatibility mode and canonical mode that could be toggled in CMake. They could ease into it with the default being compatibility, then eventually they could switch it.

Thanks for the kind words. Fortran is a great language and it's good to see efforts like this one to modernize, make more robust, and canonicalize great, but admittedly old skool, scientific software like WRF.

zbeekman commented on wrf-cmake.rb:
```diff
@@ -0,0 +1,74 @@
+
+  # Note: If you get an internal compiler error when building WRF,
+  # then you're likely running out of memory. To reduce memory usage,
+  # set HOMEBREW_MAKE_JOBS=1.
```
In core caveats are only allowed for unexpected or changed behavior as a result of packaging with homebrew, but, this information strikes me as useful enough and important enough that, while this is in a private tap, it would make sense to add this as a caveat.

zbeekman replied:
I pushed a commit making this change. We can revert it if you like.

letmaik replied:
It's great, I like it.

zbeekman committed:
wrf-cmake: Move comment about ICEs to caveat

letmaik commented on wrf-cmake.rb:
```diff
@@ -0,0 +1,77 @@
+    # Prevent CMake from finding & using ifort
+    ENV["FC"] = Formula["gcc"].opt_bin/"gfortran"
+
+    wrf_args = *std_cmake_args.map + %W[
```
Ha, one more thing! :D
Here you use `* ... .map` but for WPS it's without those. Which one is better? Whatever it is, best to make it consistent.

zbeekman replied:
That would be known as a PEBCAK error. Also known as an ID10T error, or copy pasta error. I'll push a fix.

Do you want me to squash my commits? I guess it doesn't matter, you could do it through github if it so pleases you. If you want me to, please just say the word.

letmaik replied:
I will actually keep *all* your commits. They're beautifully commented and all make sense. I'm usually too lazy and typically slip in a few "fix", "wip", "foo" commits and just squash everything at the end. But yes, your history deserves to be archived.

zbeekman committed:
wrf-cmake: Remove errant call to `map` method
